### ISSUE TITLE
Handling large numbers of messages

### ISSUE THREAD

**Issue by s-rah** [ricochet-im/ricochet#345]
Ricochet will store messages at any rate until ultimately running out of memory. Locally on my small system I can get up to ~1M messages before ricochet consistently crashes.

While certain edge cases are unavoidable, we should determine a strategy for Ricochet when there are a large number of messages in memory. 

A few possible strategies:
- Ricochet could do nothing. "This is a user problem - they should get more RAM..."
- Ricochet could limit the number of messages it accepts from each contact - and refuse newer messages.
- Ricochet could cache old messages to a file .
- Ricochet could delete old messages in favor of newer messages (once they have been read, or even before if there are enough messages)
- A combination of the above.

Likely I think the best solution is to provide a default (say 1000 messages) and to delete older messages regardless of read state - users can increase or decrease this default in the settings (most of the text messaging apps I've used have operated this way)

Relatedly, there are also UX issues when the number of messages reaches a certain threshold: The scrollbar completely vanishes making it difficult to determine where in the thread you are.

**Comment by mva1985** [comment]
As an example the surespot messaging app allows 1000 messages per conversation.  Once reached the oldest are deleted.

**Comment by mateli** [comment]
Some of us like to have our message history so deleting old messages wouldn't be a good solution. Denying messages for flood protection would be the way to go. 

Also messages should be persisted to disk using a very stable database format. Then limiting messages in ram should be done, messages should be loaded from disk on demand.

**Comment by wfr** [comment]
A simple suggestion:
https://github.com/ricochet-im/ricochet/compare/master...wfr:issue-345-limit-history?expand=1
This branch adds an option that limits the history size. Old messages are automatically pruned. The default limit is 1000 messages per conversation, as per mva1985's suggestion.

**Comment by rburchell** [comment]
@wfr Nice. I suggest opening a pull request for it :)

I'm not @special, so take my suggestions with a grain of salt, but I'm not totally sure if exposing it in the UI like that is great.

What do you think about time-based expiry rather than purely number-of-messages? Something like a combo box like this (excuse the shitty ASCII art):

    Delete messages after [ 7 days ]
                          [ 30 days ]
                          [ 30 days ]

... and as well as that, keep your existing number-of-messages deletion as a "safety-valve" of sorts, say 10k or so, some rather high number that is almost inconceivable between restarts -- as we have no search etc, I think it's fairly safe to say that if you get a 10k message history within this sort of a timespan, you probably don't care about the older messages anymore as you won't be able to find them anyway.

That "safety valve" can always be revisited & removed (and maybe a "Never" option added to the time selector) if some form of disk caching of messages is ever implemented in the future.

Then the users don't have to worry about how *much* they talk, just about how *long* they want to be able to recall history for.

**Comment by rburchell** [comment]
@mateli About persistence, if it was done, of course you are correct. But such a thing has to be considered and executed *very* carefully, and I think that such a request should be considered out of scope for the problem discussed here, which is centrally, the reliability of the software.

One of the design considerations of Ricochet is to avoid leaving metadata lying around, and logging would directly violate that, particularly worryingly is that it is not always immediately obvious - even to the people working on ricochet, let alone the end users who may not be so technically experienced - in what kinds of ways.

I believe it's things like this that have caused the caution around that so far. See #168 for more discussion on message persistence.

**Comment by special** [comment]
> Denying messages for flood protection would be the way to go.

I agree with @rburchell here. We could do more to detect contacts who are behaving in likely-malicious or definitely-malicious ways, but if we have that kind of logic, we should use it to warn the user and reject the contact, not just drop messages. Out of scope for this thread.

> What do you think about time-based expiry rather than purely number-of-messages? Something like a combo box like this (excuse the shitty ASCII art):
> ```
> Delete messages after [ 7 days ]
>                       [ 30 days ]
>                       [ 30 days ]
> ```
> Then the users don't have to worry about how much they talk, just about how long they want to be able to recall history for.

..until they close Ricochet. This UI would mislead people into thinking that Ricochet would store their last N days of messages and they would remain available. I have that concern with "Limit history" to a lesser degree also, I think because of the phrasing. Could be solved.

The time-based option is more user-friendly and familiar -- many mobile messaging clients implement it. Usually their time ranges from a few minutes to a week, and the setting is per-conversation and sometimes synced with the contact. Disappearing messages are useful -- think "I don't want to keep this on my screen / system for longer than it takes to talk about", or "I run my ricochet 24/7, and I don't want my conversations from days or weeks ago to still be there if it's compromised".

Even if it's not synchronized with contacts, this ends up slightly more complex to implement because you can't start such a short timer until the message has been seen (not just received). There should be a separate issue for this.

> ... and as well as that, keep your existing number-of-messages deletion as a "safety-valve" of sorts, say 10k or so, some rather high number

+1 to this. I'd say even 1k is enough. Ricochet's conversations are ephemeral -- if you want to keep something that happened in a past conversation, you need to actively save it. 1k is a lot of messages to scroll through.

My feeling so far is:

1. A hardcoded limit is a good idea and sufficient to solve this bug.
2. Time-based disappearing messages are a good idea, but the ideal implementation that allows short durations is non-trivial. Separate issue.
3. A time-based limit with only longer durations is potentially confusing and doesn't solve this bug any better than hardcoding does.
4. A configurable message limit is unlikely to be useful without a searchable backlog. The time-based limit solves the real use cases here better. I don't see a compelling reason to put it in the UI.

@wfr what do you think? Either way, I'd love to see a pull request :)

**Comment by wfr** [comment]
@rburchell 
> I'm not totally sure if exposing it in the UI like that is great.

I share your skepticism. One of Ricochet's major draws is simplicity and elegance.  The user won't care much if the chat window keeps the last 1000, 3333 or 10000 messages. All that matters is that the conversations are ephemeral, as @special pointed out.

>What do you think about time-based expiry rather than purely number-of-messages? 

This is a good idea. We could also consider a function that clears the chat window instantaneously. Like, a context menu in the chat window with a [Clear] option.

@special 
>+1 to this. I'd say even 1k is enough. Ricochet's conversations are ephemeral -- if you want to keep something that happened in a past conversation, you need to actively save it. 1k is a lot of messages to scroll through.

>    A hardcoded limit is a good idea and sufficient to solve this bug.
It seems we're all on the same page regarding a hardcoded limit.

I'm revising the patch with the UI removed and will post a pull request shortly.

**Comment by special** [comment]
> This is a good idea. We could also consider a function that clears the chat window instantaneously. There could be a context menu in the chat window with a [Clear] option.

#507
